test/core/test_lwt_mvar.ml

```diff
@@ -0,0 +1,90 @@
+(* OCaml promise library
+ * http://www.ocsigen.org/lwt
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, with linking exceptions;
+ * either version 2.1 of the License, or (at your option) any later
+ * version. See COPYING file for details.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ *)
+
+open Test
+open Lwt
+
+let suite =
+  suite "lwt_mvar" [
+    test "basic take"
+      (fun () ->
+         let x = Lwt_mvar.create 0 in
+         let y = Lwt_mvar.take x in
+         return (y = Lwt.return 0)
+      );
+
+    test "blocking put"
+      (fun () ->
+         let x = Lwt_mvar.create 0 in
+         let y = Lwt_mvar.put x 1 >>= (fun _ -> return 1) in
+         return (y <> Lwt.return 1)
+      );
+
+    test "put-take"
+      (fun () ->
+         let x = Lwt_mvar.create_empty () in
+         let z = Lwt_mvar.put x 0 >>= (fun _ -> Lwt_mvar.take x) in
+         return (z = Lwt.return 0)
+      );
+
+    test "take-put"
+      (fun () ->
+         let x = Lwt_mvar.create 0 in
+         let _ = Lwt_mvar.take x in
+         let z = Lwt_mvar.put x 1 >>= (fun _ -> return 2) in
+         return (z = Lwt.return 2)
+      );
+
+    test "enqueued writer"
+      (fun () ->
+         let x = Lwt_mvar.create 1 in
+         let _ = Lwt_mvar.put x 2 in
+         let y =
+           Lwt_mvar.take x >>= (fun v1 ->
+           Lwt_mvar.take x >>= (fun v2 ->
+             Lwt.return (v1 + v2))) in
+         return (y = Lwt.return 3)
+      );
+
+    test "writer cancellation"
+      (fun () ->
+         let x = Lwt_mvar.create_empty () in
+         let y = Lwt_mvar.create 1 in
+         let t, _ = Lwt.task () in
+         let r1 = t >>= (fun _ -> Lwt_mvar.put y 2) in
+         let after_cancel () =
+           let _ = Lwt_mvar.take y >>= (fun z -> Lwt_mvar.put x z); in
+           () in
+         let _ = Lwt.on_cancel r1 after_cancel in
+         Lwt.cancel r1;
```

Comment:
Changing this test should allow you to get that remaining line fully covered. Indeed, I think in this form, this test doesn't add anything beyond the tests above.

Since `t` is still pending at this point, when you cancel `r1`, it cancels `t`, and the `bind` that creates `r1` falls through `bind`'s exception-forwarding case. So the function that puts `2` to `y` is simply never called, the `put` is never even started or tried, and this is only testing how `bind` responds to `cancel`.

To trigger that line, what this test should do is get rid of `t`, and define `r1` as just that `put y 2`. It is `put y 2` that has an `on_cancel` handler added to it in `Lwt_mvar` (and we currently never call it).

 This will also be a pending promise, because `y` already has `1` at that moment. You'll want to cancel `r1`, take from `y`, check that it's `1`, take again, and check that the taker is blocked the second time.

I don't think `x` is necessary.

The reader test below has a similar issue.